corehq/apps/reports/standard/sms.py

```diff
@@ -322,51 +321,121 @@ def uses_locations(self):
         return (toggles.LOCATIONS_IN_REPORTS.enabled(self.domain)
                 and Domain.get_by_name(self.domain).uses_locations)
 
-    @property
-    def rows(self):
-        startdate = json_format_datetime(self.datespan.startdate_utc)
-        enddate = json_format_datetime(self.datespan.enddate_utc)
-        data = SMSLog.by_domain_date(self.domain, startdate, enddate)
-        result = []
-
-        reporting_locations_id = self.get_location_filter() if self.uses_locations else []
-        # Retrieve message log options
+    def _get_rows(self, paginate=True, contact_info=False):
         message_log_options = getattr(settings, "MESSAGE_LOG_OPTIONS", {})
         abbreviated_phone_number_domains = message_log_options.get("abbreviated_phone_number_domains", [])
         abbreviate_phone_number = (self.domain in abbreviated_phone_number_domains)
 
         contact_cache = {}
-        message_type_filter = self.get_message_type_filter()
 
-        for message in data:
-            if message.direction == OUTGOING and not message.processed:
-                continue
+        def filter_by_types(data_):
+            filtered_types = {t.lower() for t in MessageTypeFilter.get_value(self.request, self.domain)}
+            if not filtered_types:
+                return data_
+
+            relevant_workflows = (
+                WORKFLOW_REMINDER,
+                WORKFLOW_KEYWORD,
+                WORKFLOW_BROADCAST,
+                WORKFLOW_CALLBACK,
+                WORKFLOW_DEFAULT,
+                WORKFLOW_PERFORMANCE,
+            )
+            incl_survey = MessageTypeFilter.OPTION_SURVEY.lower() in filtered_types
+            incl_other = MessageTypeFilter.OPTION_OTHER.lower() in filtered_types
+            is_workflow_relevant = Q(workflow__in=relevant_workflows)
+            workflow_filter = Q(is_workflow_relevant & Q(workflow__in=filtered_types))
+            survey_filter = ~Q(xforms_session_couch_id__exact=None)
+            other_filter = ~Q(is_workflow_relevant | survey_filter)
+            # We can chain ANDs together, but not ORs, so we have to do all the ORs at the same time.
+            if incl_survey and incl_other:
+                filters = (workflow_filter | survey_filter | other_filter)
+            elif incl_survey:
+                filters = (workflow_filter | survey_filter)
+            elif incl_other:
+                filters = (workflow_filter | other_filter)
+            else:
+                filters = workflow_filter
+            return data_.filter(filters)
+
+        def order_by_col(data_):
+            col_fields = ['date', 'couch_recipient', 'phone_number', 'direction', 'text']
+            sort_col = self.request_params.get('iSortCol_0')
+            if sort_col is not None and sort_col < len(col_fields):
+                data_ = data_.order_by(col_fields[sort_col])
+            if self.request_params.get('sSortDir_0') == 'desc':
+                data_ = data_.reverse()
```

Comment:
It's a [QuerySet method][1], but I'm glad you pointed it out, because it needs to come with the `order_by` call. I'll check that `sSortDir_0` won't be set without `iSortCol_0`.

  [1]: https://docs.djangoproject.com/en/1.8/ref/models/querysets/#reverse